saffron/src/utils.rs

```diff
@@ -179,6 +183,41 @@ impl QueryBytes {
     }
 }
 
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[serde(bound = "F: CanonicalDeserialize + CanonicalSerialize")]
+pub struct Diff<F> {
+    #[serde_as(as = "Vec<HashMap<_, o1_utils::serialization::SerdeAs>>")]
+    pub evaluation_diffs: Vec<HashMap<usize, F>>,
```

Comment:
Note that a vector is more efficient, if you don't need to access by the modified index.

Reply:
I'm assuming that updates can be sparse enough that using a full vector here would be overkill?

Reply:
You can use a vector of type Vec<(idx, F)>, or a linear memory layout where idx and F are inlined (tuples are not, I think).